core/src/main/scala/kafka/server/metadata/KRaftMetadataCache.scala

```diff
@@ -155,72 +153,78 @@ class KRaftMetadataCache(val brokerId: Int) extends MetadataCache with Logging w
-          val partition = topic.partitions().get(partitionId)
-          val filteredReplicas = maybeFilterAliveReplicas(image, partition.replicas,
-            listenerName, false)
-          val filteredIsr = maybeFilterAliveReplicas(image, partition.isr, listenerName,
-            false)
-          val offlineReplicas = getOfflineReplicas(image, partition, listenerName)
-          val maybeLeader = getAliveEndpoint(image, partition.leader, listenerName)
-          maybeLeader match {
-            case None =>
-              val error = if (!image.cluster().brokers.containsKey(partition.leader)) {
-                debug(s"Error while fetching metadata for $topicName-$partitionId: leader not available")
-                Errors.LEADER_NOT_AVAILABLE
-              } else {
-                debug(s"Error while fetching metadata for $topicName-$partitionId: listener $listenerName " +
-                  s"not found on leader ${partition.leader}")
-                Errors.LISTENER_NOT_FOUND
-              }
-              result.addOne(new DescribeTopicPartitionsResponsePartition()
-                .setErrorCode(error.code)
-                .setPartitionIndex(partitionId)
-                .setLeaderId(MetadataResponse.NO_LEADER_ID)
-                .setLeaderEpoch(partition.leaderEpoch)
-                .setReplicaNodes(filteredReplicas)
-                .setIsrNodes(filteredIsr)
-                .setOfflineReplicas(offlineReplicas))
-            case Some(leader) =>
-              val error = if (filteredReplicas.size < partition.replicas.length) {
-                debug(s"Error while fetching metadata for $topicName-$partitionId: replica information not available for " +
-                  s"following brokers ${partition.replicas.filterNot(filteredReplicas.contains).mkString(",")}")
-                Errors.REPLICA_NOT_AVAILABLE
-              } else if (filteredIsr.size < partition.isr.length) {
-                debug(s"Error while fetching metadata for $topicName-$partitionId: in sync replica information not available for " +
-                  s"following brokers ${partition.isr.filterNot(filteredIsr.contains).mkString(",")}")
-                Errors.REPLICA_NOT_AVAILABLE
-              } else {
-                Errors.NONE
-              }
-
-              result.addOne(new DescribeTopicPartitionsResponsePartition()
-                .setErrorCode(error.code)
-                .setPartitionIndex(partitionId)
-                .setLeaderId(leader.id())
-                .setLeaderEpoch(partition.leaderEpoch)
-                .setReplicaNodes(filteredReplicas)
-                .setIsrNodes(filteredIsr)
-                .setOfflineReplicas(offlineReplicas)
-                .setEligibleLeaderReplicas(Replicas.toList(partition.elr))
-                .setLastKnownElr(Replicas.toList(partition.lastKnownElr)))
+        // The partition id may not be consecutive.
+        val partitions = topic.partitions().keySet().stream().sorted().iterator()
```

Comment:
This has O(N*logN) runtime complexity and O(N) space complexity.  We could do O(N) complexity and not have an extra copy if we just iterate over all partitions and filter the ones that fit into the required range (one of your previous implementations had this).

Reply:
I am not sure I get it. The partition IDs can be random like the cases in UT, I don't have an O(n) with no extra space simple solution off the top of my head. Maybe running the quick select can do the trick but it is not generically supported by Java.
Instead, I use a tree set to maintain the top K smallest partitions larger than the start index. This is better than the original sorting.